sdk/core/azure-core/test/ut/curl_session_test.cpp

```diff
@@ -78,4 +80,33 @@ namespace Azure { namespace Core { namespace Test {
     // Clear the connections from the pool to invoke clean routine
     Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.clear();
   }
+
+  TEST_F(CurlSession, DoNotReuseConnectionIfDownloadFail)
+  {
+
+    // Can't mock the curMock directly from a unique ptr, heap allocate it first and then make a
+    // unique ptr for it
+    MockCurlNetworkConnection* curlMock = new MockCurlNetworkConnection();
+    // mock an upload error
+    EXPECT_CALL(*curlMock, SendBuffer(_, _, _)).WillOnce(Return(CURLE_SEND_ERROR));
+    EXPECT_CALL(*curlMock, DestructObj());
+
+    // Create the unique ptr to take care about memory free at the end
+    std::unique_ptr<MockCurlNetworkConnection> uniqueCurlMock(curlMock);
+
+    // Simulate a request to be sent
+    Azure::Core::Http::Url url("http://microsoft.com");
+    Azure::Core::Http::Request request(Azure::Core::Http::HttpMethod::Get, url);
+
+    {
+      // Create the session inside scope so it is released and the connection is moved to the pool
+      auto session = std::make_unique<Azure::Core::Http::CurlSession>(
+          request, std::move(uniqueCurlMock), true);
+
+      auto returnCode = session->Perform(Azure::Core::GetApplicationContext());
+      EXPECT_EQ(CURLE_SEND_ERROR, returnCode);
+    }
+    // Check connection pool is empty (connection was not moved to the pool)
+    EXPECT_EQ(Azure::Core::Http::CurlConnectionPool::ConnectionPoolIndex.size(), 0);
```

Comment:
Just to confirm, will this test fail if we revert the code change from https://github.com/Azure/azure-sdk-for-cpp/pull/863?

Reply:
Yes, the upload would fail and isEOF() would be `true` (because the totalRead ==0 && content_lenght ==0).
Content length was never set.
Hence, destructing the session would think the EOF was reached and move the connection to the pool.
With the fix, the also check if we ever started to stream data (start reading from the wire)